### PR TITLE
Allow optionally aggregating all fixes into one pull request

### PR DESCRIPTION
**Pull request by EyalDelarea**
- [x] All [tests](https://github.com/jfrog/frogbot#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
---

**Adds the ability to aggregate pull requests**

- Only one pull request will be active at any given time.
- Frogbot will force push to the existing branch in case of an already open Pull Request
- Can be set from ENV var or Config file.

Aggregate Fixes Flow:
image

Update pull request:
image

### CODE REVIEW

File: commands/utils/git.go [hunk 1]
```diff
@@ -238,6 +242,14 @@ func (gm *GitManager) GenerateCommitMessage(impactedPackage string, fixVersion s
 	return formatStringWithPlaceHolders(template, impactedPackage, fixVersion, "", true)
 }
 
+func (gm *GitManager) GenerateAggregatedCommitMessage() string {
```
Comment (eyalbe4) [inline commands/utils/git.go:242]: We shouod add a unit test for this function, similar to the unit test we have for `GenerateCommitMessage`.

File: commands/utils/git.go [hunk 2]
```diff
@@ -280,6 +292,60 @@ func (gm *GitManager) GeneratePullRequestTitle(impactedPackage string, version s
...
+		// Delete outdated Frogbot's branches
+		if isSafeToDelete(ignoreBranchName, deleteCandidateBranch, author) {
+			err = refList.Push(&git.PushOptions{
+				RemoteName: gm.remoteName,
+				RefSpecs:   []config.RefSpec{config.RefSpec(":refs/heads/" + deleteCandidateBranch)},
+				Auth:       gm.auth,
+			})
+			if err != nil {
+				log.Warn("Failed to delete branch:", deleteCandidateBranch)
+				return err
```
Comment (eyalbe4) [inline commands/utils/git.go]: The above two lines can be replaced with - 
```
return fmt.Errorf("%s '%s'. Cause: %s", "Failed to delete branch", deleteCandidateBranch, err.Error())
```

File: commands/utils/params.go [hunk 3]
```diff
@@ -290,7 +292,16 @@ func extractGitParamsFromEnv() (*Git, error) {
 	gitParams.Branches = append(gitParams.Branches, branch)
 	if pullRequestIDString := getTrimmedEnv(GitPullRequestIDEnv); pullRequestIDString != "" {
 		gitParams.PullRequestID, err = strconv.Atoi(pullRequestIDString)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Non-mandatory git aggregate pull requests feature
```
Comment (eyalbe4) [inline commands/utils/params.go:299]: Non-mandatory git aggregate pull requests feature
-->
Non-mandatory pull requests aggregation option

File: commands/createfixpullrequests.go [hunk 4]
```diff
@@ -30,6 +30,8 @@ type CreateFixPullRequestsCmd struct {
 	projectWorkingDir string
 	// The git client the command performs git operations with
 	gitManager *utils.GitManager
+	// Decide if to fix each vulnerability separately or to aggregate to one pull request at any given time
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:33]: // Decide if to fix each vulnerability separately or to aggregate to one pull request at any given time
-->
// Determines whether to open a pull request for each vulnerability fix or to aggregate all fixes into one pull request.

File: commands/createfixpullrequests.go [hunk 5]
```diff
@@ -133,9 +135,17 @@ func (cfp *CreateFixPullRequestsCmd) fixVulnerablePackages(fixVersionsMap map[st
...
-	// Fix all impacted packages
+	if cfp.aggregatePullRequests {
+		err = cfp.fixPackagesAggregated(fixVersionsMap)
+	} else {
+		err = cfp.fixPackagesSeparately(fixVersionsMap)
+	}
+	return
+}
+
+func (cfp *CreateFixPullRequestsCmd) fixPackagesSeparately(fixVersionsMap map[string]*utils.FixVersionInfo) (err error) {
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:147]: fixPackagesSeparately --> fixIssuesSeparatePRs

File: commands/createfixpullrequests.go [hunk 6]
```diff
@@ -147,10 +157,46 @@ func (cfp *CreateFixPullRequestsCmd) fixVulnerablePackages(fixVersionsMap map[st
 	return
 }
 
-func (cfp *CreateFixPullRequestsCmd) fixSinglePackage(impactedPackage string, fixVersionInfo *utils.FixVersionInfo) (err error) {
+func (cfp *CreateFixPullRequestsCmd) fixPackagesAggregated(fixVersionsMap map[string]*utils.FixVersionInfo) (err error) {
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:161]: fixPackagesAggregated --> fixIssuesSinglePR

File: commands/utils/params.go [hunk 7]
```diff
@@ -129,6 +129,7 @@ type Git struct {
 	BranchNameTemplate       string   `yaml:"branchNameTemplate,omitempty"`
 	CommitMessageTemplate    string   `yaml:"commitMessageTemplate,omitempty"`
 	PullRequestTitleTemplate string   `yaml:"pullRequestTitleTemplate,omitempty"`
+	AggregatePullRequests    bool     `yaml:"aggregatePullRequests,omitempty"`
```
Comment (eyalbe4) [inline commands/utils/params.go:132]: AggregatePullRequests --> AggregateFixes
aggregatePullRequests --> aggregateFixes

File: commands/utils/params.go [hunk 8]
```diff
@@ -290,7 +292,16 @@ func extractGitParamsFromEnv() (*Git, error) {
 	gitParams.Branches = append(gitParams.Branches, branch)
 	if pullRequestIDString := getTrimmedEnv(GitPullRequestIDEnv); pullRequestIDString != "" {
 		gitParams.PullRequestID, err = strconv.Atoi(pullRequestIDString)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Non-mandatory git aggregate pull requests feature
+	aggregatePullRequestsEnv, err := getBoolEnv(GitAggregatePullRequestsEnv, false)
```
Comment (eyalbe4) [inline commands/utils/params.go:300]: aggregatePullRequestsEnv --> aggregateFixesEnv
(let's change all functions and related names throughout the code accordingly).

File: docs/templates/.frogbot/frogbot-config.yml [hunk 9]
```diff
@@ -29,6 +29,10 @@
       # The template can optionally include the ${IMPACTED_PACKAGE} and ${FIX_VERSION} variables.
       # pullRequestTitleTemplate: "[🐸 Frogbot] Upgrade ${IMPACTED_PACKAGE} to to ${FIX_VERSION}"
 
+      # [Default: false]
+      # Frogbot will aggregate all vulnerabilities fixes to one pull request at any given time
+      # aggregatePullRequests: false
```
Comment (eyalbe4) [inline docs/templates/.frogbot/frogbot-config.yml:34]: I recomment changing to - 
```
      # [Default: false]
      # If true, Frogbot creates a single pull request with all the fixes.
      $ If fase, Frogbot creates a separate pull request for each fix.
      # aggregateFixes: false
```
Let's also add this variable (JF_AGGREGATE_FIXES) to all workflow docs.

File: commands/createfixpullrequests.go [hunk 10]
```diff
@@ -195,22 +237,52 @@ func (cfp *CreateFixPullRequestsCmd) openFixingPullRequest(impactedPackage, fixB
...
+	}
+	log.Info("Pushing fix branch:", fixBranchName, "...")
+	err = cfp.gitManager.Push(true)
 	if err != nil {
 		return
 	}
+	if !exists {
+		pullRequestTitle := utils.AggregatedTitleTemplate
+		log.Info("Creating Pull Request form:", fixBranchName, " to:", cfp.details.Branch)
+		prBody := commitMessage + "\n\n" + utils.WhatIsFrogbotMd
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:269]: I suggest we leave the PR body empty for now, till we decide what its format should be.

File: commands/createfixpullrequests.go [hunk 11]
```diff
@@ -309,6 +366,21 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(impactedPackage
 	return packageHandler.UpdateImpactedPackage(impactedPackage, fixVersionInfo)
 }
 
+// getMinimalFixVersion that fixes the current impactedPackage
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:369]: // getMinimalFixVersion that fixes the current impactedPackage
-->
// getMinimalFixVersion find the minimal version that fixes the current impactedPackage;

File: commands/createfixpullrequests.go [hunk 12]
```diff
@@ -309,6 +366,21 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(impactedPackage
...
+// fixVersions array is sorted, so we take the first index, unless it's version is older than what we have now.
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:370]: // fixVersions array is sorted, so we take the first index, unless it's version is older than what we have now.
-->
// fixVersions is a sorted array. The function returns the first version in the array, that is larger than `impactedPackageVersion`.

File: docs/install-azure-repos.md [hunk 13]
```diff
@@ -143,6 +143,11 @@ To install Frogbot on Azure Repos repositories, follow these steps.
             # Template for the pull request title generated by Frogbot when creating pull requests with fixes.
             # The template can optionally include the ${IMPACTED_PACKAGE} and ${FIX_VERSION} variables.
             # JF_PULL_REQUEST_TITLE_TEMPLATE: "[🐸 Frogbot] Upgrade ${IMPACTED_PACKAGE} to to ${FIX_VERSION}"
+
+            # [Optional, Default: "FALSE"]
+            # If true, Frogbot creates a single pull request with all the fixes.
+            # If false, Frogbot creates a separate pull request for each fix.
```
Comment (eyalbe4) [inline docs/install-azure-repos.md:149]: true --> TRUE
false --> FALSE
(unlike in the frogbot-config.yml when the above values should be documented as lower-case, here it should be documented as upper-case, beucase of the way this value is set in the line below. Let's implement the same in all workflows).

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: commands/createfixpullrequests.go (hunk 10 above)
Comment (EyalDelarea) [reply commands/createfixpullrequests.go:269]: The body now is the commit message + link to Frogbot as it was, Do we want to be empty ?

File: commands/createfixpullrequests.go [hunk 14]
```diff
@@ -30,6 +30,8 @@ type CreateFixPullRequestsCmd struct {
 	projectWorkingDir string
 	// The git client the command performs git operations with
 	gitManager *utils.GitManager
+	// Determines whether to open a pull request for each vulnerability fix or to aggregate all fixes into one pull request.
+	aggregatePullRequests bool
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:34]: aggregatePullRequests --> aggregateFixes

File: commands/createfixpullrequests.go [hunk 15]
```diff
@@ -178,14 +220,12 @@ func (cfp *CreateFixPullRequestsCmd) openFixingPullRequest(impactedPackage, fixB
 
 	commitMessage := cfp.gitManager.GenerateCommitMessage(impactedPackage, fixVersionInfo.FixVersion)
 	log.Info("Running git add all and commit...")
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:222]: Running git add all and commit... --> Running 'git add --all' and 'git commit'...

Let's add the same change in the `openAggregatedPullRequest` function.

File: commands/createfixpullrequests.go [hunk 16]
```diff
@@ -178,14 +220,12 @@ func (cfp *CreateFixPullRequestsCmd) openFixingPullRequest(impactedPackage, fixB
...
-	err = cfp.gitManager.AddAllAndCommit(commitMessage)
-	if err != nil {
+	if err = cfp.gitManager.AddAllAndCommit(commitMessage); err != nil {
 		return
 	}
 
 	log.Info("Pushing fix branch:", fixBranchName, "...")
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:227]: Pushing fix branch: --> Pushing branch:

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request merged**